File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        l=[]
        for i in range(len(nums)):
            if i > 0 and nums[i]==nums[i-1]:
                continue 

            j=i+1
            k=len(nums)-1
            while j < k:

                c=nums[i] + nums[j] + nums[k]
                if c==0:
                    l.append([nums[i] , nums[j] , nums[k]])
                    j+=1
                    while  nums[j]==nums[j-1] and j< k:
                        j+=1
                    
                elif c > 0:
                    k-=1
                else :
                    j+=1
        return l


        

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int l,r,sum;
        int n=nums.length;
        for (int j = 0 ; j <n ;j++ ){
            if (j > 0 && nums[j-1]==nums[j]){
                continue;

            }
            l=j+1;
            r=n-1;
            while(l < r){
                sum=nums[j]+nums[l]+nums[r];
                if (sum==0){
                    List<Integer> ele =Arrays.asList(nums[j],nums[l],nums[r]);
                    result.add(ele);
                    l++;
                    r--;
                    while (l < n &&nums[l] == nums[l-1]) l++;
                    while ( r > 0 && nums[r]==nums[r+1]) r--;
                   
                    
                }
                else if(sum > 0) r--;

                else l++;
            }
                    
                    
                
            
        }
        return result;
        
    }
}